source/server/callback_router.h

```diff
@@ -0,0 +1,112 @@
+#ifndef NIDEVICE_GRPC_DEVICE_CALLBACK_ROUTER_H
+#define NIDEVICE_GRPC_DEVICE_CALLBACK_ROUTER_H
+
+#include <atomic>
+#include <functional>
+#include <mutex>
+#include <unordered_map>
+
+namespace nidevice_grpc {
+
+// Allows registering Handlers taking TArgs... and exposes a static handle_callback
+// method that can be passed as a function pointer to C APIs.
+// This allows Handlers to be implemented with stateful lambdas, which are not convertible
+// to function pointers.
+template <typename TReturn, typename... TArgs>
+class CallbackRouter {
+ public:
+  using Handler = std::function<TReturn(TArgs...)>;
+  using Token = void*;
+
+  // Move-only RAII type exposed through register_handler, to auto-unregister_handler when leaving scope.
+  class CallbackRegistration {
+   public:
+    CallbackRegistration() : token_(0) {}
+    CallbackRegistration(Token token) : token_(token) {}
+    CallbackRegistration(const CallbackRegistration&) = delete;
+    CallbackRegistration& operator=(const CallbackRegistration&) = delete;
+    CallbackRegistration(CallbackRegistration&& rhs) : token_(rhs.token_) { rhs.token_ = reinterpret_cast<Token>(0); }
```

Comment:
This needs to be move-able to allow returning from `register_handler` because the version of gcc we're using on linux RT doesn't implement "required move elision" (I think).

The return from that method will likely change to a `unique_ptr` to support the async implementation. In that case, this wouldn't be needed anyway.